Common/src/main/java/com/blamejared/crafttweaker/impl/registry/natives/ExecutableReferenceInfo.java

```diff
@@ -0,0 +1,158 @@
+                if(result.containsKey(k)) {
+                    throw new IllegalStateException("Attempted to add the same annotation class " + k + " more than once");
+                }
+                result.put(k, v);
+            });
+            return result;
+        }
+        
+        private ZenCodeType.Method createMethodAnnotation() {
+            
+            return new ZenCodeType.Method() {
+                @Override
+                public String value() {
+                    
+                    return "";
+                }
+                
+                @Override
+                public Class<? extends Annotation> annotationType() {
+                    
+                    return ZenCodeType.Method.class;
+                }
+            };
+        }
+        
+        private ZenCodeType.Constructor createConstructorAnnotation() {
+            
+            return new ZenCodeType.Constructor() {
+                @Override
+                public Class<? extends Annotation> annotationType() {
+                    
+                    return ZenCodeType.Constructor.class;
+                }
+            };
+        }
+        
+        private ZenCodeType.Getter createGetterAnnotation(final String setterName) {
+            
+            return new ZenCodeType.Getter() {
+                @Override
+                public String value() {
+                    
+                    return setterName;
+                }
+                
+                @Override
+                public Class<? extends Annotation> annotationType() {
+                    
+                    return ZenCodeType.Getter.class;
+                }
+            };
+        }
+        
+        private ZenCodeType.Setter createSetterAnnotation(final String setterName) {
+            
+            return new ZenCodeType.Setter() {
+                @Override
+                public String value() {
+                    
+                    return setterName;
+                }
+                
+                @Override
+                public Class<? extends Annotation> annotationType() {
+                    
+                    return ZenCodeType.Setter.class;
+                }
+            };
+        }
+        
+    }
+    
+    private final List<Class<?>> arguments;
+    private final Map<Class<? extends Annotation>, Annotation> presentAnnotationTypes;
+    
+    ExecutableReferenceInfo(final List<Class<?>> arguments, final Map<Class<? extends Annotation>, Annotation> presentAnnotationTypes) {
+        
+        this.arguments = Collections.unmodifiableList(arguments);
+        this.presentAnnotationTypes = Collections.unmodifiableMap(presentAnnotationTypes);
+    }
+    
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Annotation> Optional<T> getAnnotation(final Class<T> annotationClass) {
+        
+        return Optional.ofNullable(this.presentAnnotationTypes.get(annotationClass)).map(it -> (T) it);
```

Comment:
Why not cast here using the reflection info?
This way we would error here and not type-error wherever the optional is used later if something went wrong

```suggestion
        return Optional.ofNullable(this.presentAnnotationTypes.get(annotationClass)).map(annotationClass::cast);
```